tests/testsync.nim

```diff
@@ -41,6 +41,75 @@ suite "Asynchronous sync primitives test suite":
       poll()
     result = testLockResult
 
+  proc testFlag(): Future[bool] {.async.} =
+    var lock = newAsyncLock()
+    var futs: array[4, Future[void]]
+    futs[0] = lock.acquire()
+    futs[1] = lock.acquire()
+    futs[2] = lock.acquire()
+    futs[3] = lock.acquire()
+
+    proc checkFlags(b0, b1, b2, b3, b4: bool): bool =
+      (lock.locked == b0) and
+        (futs[0].finished == b1) and (futs[1].finished == b2) and
+        (futs[2].finished == b3) and (futs[3].finished == b4)
+
+    if not(checkFlags(true, true, false, false ,false)):
+      return false
+
+    lock.release()
+    if not(checkFlags(true, true, false, false, false)):
+      return false
+    await sleepAsync(10.milliseconds)
+    if not(checkFlags(true, true, true, false, false)):
+      return false
+
+    lock.release()
+    if not(checkFlags(true, true, true, false, false)):
+      return false
+    await sleepAsync(10.milliseconds)
+    if not(checkFlags(true, true, true, true, false)):
+      return false
+
+    lock.release()
+    if not(checkFlags(true, true, true, true, false)):
+      return false
+    await sleepAsync(10.milliseconds)
+    if not(checkFlags(true, true, true, true, true)):
+      return false
+
+    lock.release()
+    if not(checkFlags(false, true, true, true, true)):
+      return false
+    await sleepAsync(10.milliseconds)
+    if not(checkFlags(false, true, true, true, true)):
+      return false
+
+    return true
+
+  proc testNoAcquiredRelease(): Future[bool] {.async.} =
+    var lock = newAsyncLock()
+    var res = false
+    try:
+      lock.release()
+    except AsyncLockError:
+      res = true
+    return res
+
+  proc testDoubleRelease(): Future[bool] {.async.} =
+    var lock = newAsyncLock()
+    var fut0 = lock.acquire()
+    var fut1 = lock.acquire()
+    var res = false
+    asyncSpawn fut0
+    asyncSpawn fut1
```

Comment:
this depends on `asyncSpawn` not yielding control - is that the intent? is that a guarantee that will be maintained?

Reply:
`asyncSpawn` here is to raise `Defect` on if `fut0` or `fut1` will be cancelled or failed.